### PR TITLE
SWC import modularization plugin

### PR DESCRIPTION
**Pull request by losfair**
## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [x] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [ ] Make sure the linting passes by running `yarn lint`

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| buildDuration | 19.2s | 20s | ⚠️ +783ms |
| buildDurationCached | 7.8s | 8.3s | ⚠️ +566ms |
| nodeModulesSize | 383 MB | 383 MB | ⚠️ +10 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.748 | 3.824 | ⚠️ +0.08 |
| / avg req/sec | 667.02 | 653.72 | ⚠️ -13.3 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.665 | 1.645 | -0.02 |
| /error-in-render avg req/sec | 1501.8 | 1519.93 | +18.13 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 27.9 kB | 27.9 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.5 kB | 71.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 326 B | 326 B | ✓ |
| dynamic-HASH.js gzip | 2.57 kB | 2.57 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 919 B | 919 B | ✓ |
| image-HASH.js gzip | 5.05 kB | 5.05 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.26 kB | 2.26 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.7 kB | 14.7 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| buildDuration | 24.4s | 23.2s | -1.2s |
| buildDurationCached | 8s | 7.4s | -609ms |
| nodeModulesSize | 383 MB | 383 MB | ⚠️ +10 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.741 | 3.782 | ⚠️ +0.04 |
| / avg req/sec | 668.19 | 661.05 | ⚠️ -7.14 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.657 | 1.646 | -0.01 |
| /error-in-render avg req/sec | 1508.79 | 1519.02 | +10.23 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.1 kB | 28.1 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.1 kB | 72.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 326 B | 326 B | ✓ |
| dynamic-HASH.js gzip | 2.56 kB | 2.56 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.2 kB | 5.2 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.33 kB | 2.33 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.9 kB | 14.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| _buildManifest.js gzip | 456 B | 456 B | ✓ |
| Overall change | 456 B | 456 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| index.html gzip | 530 B | 530 B | ✓ |
| link.html gzip | 543 B | 543 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: 84b5fb2b51824361784a713513ece86381363662

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 869148f3957d113962c4575b63692cbfa5d15b3b

`yarn testheadless test/integration/async-modules/test/index.test.js` 
- Async modules > production mode > can render async AMP pages

<details>
<summary>Expand output</summary>

  ● Async modules › production mode › can render async AMP pages

    TIMED OUT: just now

    11 seconds ago

      474 |
      475 |   if (hardError) {
    > 476 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content)
          |           ^
      477 |   }
      478 |   return false
      479 | }

      at Object.check (lib/next-test-utils.js:476:11)
      at Object.<anonymous> (integration/async-modules/test/index.test.js:85:7)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by vinnieking06** [comment]
Our team is currently migrating away from babel to swc, so we're excited to bring this transform-import functionality into our codebase. Any update on when this will be released?

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 7cbc423b7ab6e38100777c4695df7f7bdb55838b

`yarn testheadless test/production/next/jest/index.test.ts` 
- next/jest > should work

<details>
<summary>Expand output</summary>

  ● next/jest › should work

    thrown: "Exceeded timeout of 90000 ms for a hook.
    Use jest.setTimeout(newTimeout) to increase the timeout value, if this is a long-running test."

       6 |   let next: NextInstance
       7 |
    >  8 |   beforeAll(async () => {
         |   ^
       9 |     next = await createNext({
      10 |       files: {
      11 |         'public/vercel.svg':

      at production/next/jest/index.test.ts:8:3
      at Object.<anonymous> (production/next/jest/index.test.ts:5:1)


  ● Test suite failed to run

    TypeError: Cannot read property 'destroy' of undefined

      134 |     })
      135 |   })
    > 136 |   afterAll(() => next.destroy())
          |                       ^
      137 |
      138 |   it('should work', async () => {
      139 |     const html = await renderViaHTTP(next.url, '/')

      at production/next/jest/index.test.ts:136:23

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| buildDuration | 19.3s | 19.6s | ⚠️ +241ms |
| buildDurationCached | 7.7s | 7.4s | -224ms |
| nodeModulesSize | 467 MB | 467 MB | ⚠️ +813 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.108 | 4.234 | ⚠️ +0.13 |
| / avg req/sec | 608.52 | 590.53 | ⚠️ -17.99 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 2.231 | 2.11 | -0.12 |
| /error-in-render avg req/sec | 1120.59 | 1185.05 | +64.46 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28 kB | 28 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.6 kB | 71.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 309 B | 309 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.57 kB | 2.57 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.48 kB | 5.48 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.26 kB | 2.26 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 387 B | 387 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.2 kB | 15.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| buildDuration | 23.9s | 23.8s | -45ms |
| buildDurationCached | 7.4s | 7.5s | ⚠️ +9ms |
| nodeModulesSize | 467 MB | 467 MB | ⚠️ +813 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.243 | 4.228 | -0.02 |
| / avg req/sec | 589.15 | 591.36 | +2.21 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 2.218 | 2.132 | -0.09 |
| /error-in-render avg req/sec | 1127 | 1172.65 | +45.65 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.2 kB | 28.2 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.1 kB | 72.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 324 B | 324 B | ✓ |
| dynamic-HASH.js gzip | 2.56 kB | 2.56 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.59 kB | 5.59 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.33 kB | 2.33 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 388 B | 388 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.3 kB | 15.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: 7cbc423b7ab6e38100777c4695df7f7bdb55838b

File: packages/next-swc/crates/core/src/modularize_imports.rs [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        .renderer
+        .register_helper("camelCase", Box::new(helper_camel_case));
+    for (mut k, v) in config.packages {
+        // XXX: Should we keep this hack?
+        if !k.starts_with('^') && !k.ends_with('$') {
+            k = format!("^{}$", k);
+        }
+        folder
+            .packages
+            .push((Regex::new(&k).expect("transform-imports: invalid regex"), v));
```
Comment (kdy1) [inline packages/next-swc/crates/core/src/modularize_imports.rs:184]: This will be slow, because `Regex` will be created once per file.

Comment (kdy1) [reply packages/next-swc/crates/core/src/modularize_imports.rs:184]: You can use https://docs.rs/swc_cached/0.1.1/swc_cached/regex/struct.CachedRegex.html

File: packages/next-swc/crates/core/src/modularize_imports.rs [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    pub transform: String,
+    #[serde(default)]
+    pub prevent_full_import: bool,
+    #[serde(default)]
+    pub skip_default_conversion: bool,
+}
+
+struct FoldImports {
+    renderer: handlebars::Handlebars<'static>,
+    packages: Vec<(Regex, PackageConfig)>,
```
Comment (kdy1) [inline packages/next-swc/crates/core/src/modularize_imports.rs:31]: You can change this to `Vec<(CachedRegex, PackageConfig)>`

**Review by kdy1 — CHANGES_REQUESTED** [review]
Nice work!
I have one performance concern

File: packages/next-swc/crates/core/src/modularize_imports.rs (hunk 1 above)
Comment (Brooooooklyn) [reply packages/next-swc/crates/core/src/modularize_imports.rs:184]: Or use `once_cell::Lazy` https://github.com/vercel/next.js/blob/canary/packages/next-swc/crates/core/src/emotion/mod.rs#L75

File: packages/next-swc/crates/core/src/modularize_imports.rs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    group: Vec<&'a str>,
+}
+
+impl<'a> Rewriter<'a> {
+    fn rewrite(&self, old_decl: &ImportDecl) -> Vec<ImportDecl> {
+        if old_decl.type_only || old_decl.asserts.is_some() {
+            return vec![old_decl.clone()];
+        }
+
+        let mut out: Vec<ImportDecl> = vec![];
```
Comment (Brooooooklyn) [inline packages/next-swc/crates/core/src/modularize_imports.rs:47]: `Vec::with_capacity` to prevent reallocation

**Review by kdy1 — DISMISSED** [review]
LGTM.

**Review by Brooooooklyn — DISMISSED** [review]

**Comment by ijjk** [comment]
## Failing test suites 
Commit: f56478a237a339bb25dcd4557c7d924ebae501b7

`yarn testheadless test/production/next/jest/index.test.ts` 
- next/jest > should work

<details>
<summary>Expand output</summary>

  ● next/jest › should work

    thrown: "Exceeded timeout of 90000 ms for a hook.
    Use jest.setTimeout(newTimeout) to increase the timeout value, if this is a long-running test."

       6 |   let next: NextInstance
       7 |
    >  8 |   beforeAll(async () => {
         |   ^
       9 |     next = await createNext({
      10 |       files: {
      11 |         'public/vercel.svg':

      at production/next/jest/index.test.ts:8:3
      at Object.<anonymous> (production/next/jest/index.test.ts:5:1)


  ● Test suite failed to run

    TypeError: Cannot read property 'destroy' of undefined

      134 |     })
      135 |   })
    > 136 |   afterAll(() => next.destroy())
          |                       ^
      137 |
      138 |   it('should work', async () => {
      139 |     const html = await renderViaHTTP(next.url, '/')

      at production/next/jest/index.test.ts:136:23

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| buildDuration | 15.4s | 15.3s | -120ms |
| buildDurationCached | 6.1s | 6s | -116ms |
| nodeModulesSize | 467 MB | 467 MB | ⚠️ +813 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.935 | 2.97 | ⚠️ +0.04 |
| / avg req/sec | 851.9 | 841.75 | ⚠️ -10.15 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.236 | 1.208 | -0.03 |
| /error-in-render avg req/sec | 2021.87 | 2069.01 | +47.14 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28 kB | 28 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.6 kB | 71.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 309 B | 309 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.57 kB | 2.57 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.48 kB | 5.48 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.26 kB | 2.26 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 387 B | 387 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.2 kB | 15.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| buildDuration | 18.8s | 19.1s | ⚠️ +298ms |
| buildDurationCached | 6.2s | 6.1s | -91ms |
| nodeModulesSize | 467 MB | 467 MB | ⚠️ +813 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.97 | 3.015 | ⚠️ +0.04 |
| / avg req/sec | 841.78 | 829.15 | ⚠️ -12.63 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.197 | 1.193 | 0 |
| /error-in-render avg req/sec | 2088.99 | 2095.63 | +6.64 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.2 kB | 28.2 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.1 kB | 72.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 324 B | 324 B | ✓ |
| dynamic-HASH.js gzip | 2.56 kB | 2.56 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.59 kB | 5.59 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.33 kB | 2.33 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 388 B | 388 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 15.3 kB | 15.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | losfair/next.js feature/transform-imports | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: 7acf9a309ab3f991f85ac0f8fe34eb3bfa89f9db

**Review by Brooooooklyn — APPROVED** [review]

**Review by timneutkens — APPROVED** [review]

**Pull request merged**